### PR TITLE
Add last_compaction_state to sys.segments table

### PR DESCRIPTION
**Pull request by maytasm**
Add last_compaction_state to sys.segments table

### Description

Segments being compacted or not compacted have a big impact on query performance. 
Currently it is not easy to verify questions like:
- if performance of query can be improve by compaction 
- if a performance of query is bad, was it because segments for that interval was not compacted.
- if a segment was compacted, what was it compacted with? was it compacted with the auto compaction config set?

This PR expose the lastCompactionState via the sys.segments table (so user can do query with slice/dice, group, time filter, etc) by using the lastCompactionState stored within the segment metadata. If a compactionState exist then the segment was compacted. This PR also change storing lastCompactionState as default for manual compaction in addition to auto compaction.

For example, if you want to know if the segment was compacted or not (ANY compaction), you can do `WHERE last_compaction_state is not null` and if you want to know if the segment was compacted with current auto compaction configs, you can do `WHERE last_compaction_state == '{current_auto_compact_config}'` 

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: core/src/main/java/org/apache/druid/timeline/DataSegment.java [hunk 1]
```diff
@@ -100,7 +100,7 @@
   /**
    * Stores some configurations of the compaction task which created this segment.
    * This field is filled in the metadata store only when "storeCompactionState" is set true in the context of the
-   * compaction task which is false by default.
+   * compaction task (true by default).
```
Comment (suneet-s) [inline core/src/main/java/org/apache/druid/timeline/DataSegment.java:103]: super nit - to link to where the default is.
```suggestion
   * compaction task (true by default). {@link org.apache.druid.indexing.common.task.Tasks#DEFAULT_STORE_COMPACTION_STATE}.
```

File: docs/querying/sql.md [hunk 2]
```diff
@@ -1086,6 +1086,7 @@ Segments table provides details on all Druid segments, whether they are publishe
 |shardSpec|STRING|The toString of specific `ShardSpec`|
 |dimensions|STRING|The dimensions of the segment|
 |metrics|STRING|The metrics of the segment|
+|last_compaction_state|STRING|The configurations of the compaction task which created this segment. May be null if segment was not created by compaction task.|
 
 For example to retrieve all segments for datasource "wikipedia", use the query:
```
Comment (suneet-s) [inline docs/querying/sql.md:1091]: The PR description talks about being able to identify performance issues where looking at last_compaction_state can help. Should we add some example sql statements below to explain how to identify the 3 issues outlined in the PR description?

**Comment by suneet-s** [comment]
Added Release Notes because it changed the default behavior of `Tasks.DEFAULT_STORE_COMPACTION_STATE`

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java [hunk 3]
```diff
@@ -311,7 +312,8 @@ public TableType getJdbcTableType()
                 val.isOvershadowed() ? IS_OVERSHADOWED_TRUE : IS_OVERSHADOWED_FALSE,
                 segment.getShardSpec(),
                 segment.getDimensions(),
-                segment.getMetrics()
+                segment.getMetrics(),
+                segment.getLastCompactionState()
```
Comment (suneet-s) [inline sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:316]: `CompactionState` includes the `PartitionsSpec` and `indexSpec`. 

What should a caller of the API do with these?
The compaction state appears to be an unbounded json object. Do we foresee any issues around serializing the entire json blob every time we need to call the sys table?

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java [hunk 4]
```diff
@@ -422,6 +422,7 @@ ParallelIndexSupervisorTask newTask(String taskId, ParallelIndexIngestionSpec in
   {
     final Map<String, Object> newContext = new HashMap<>(getContext());
     newContext.put(CTX_KEY_APPENDERATOR_TRACKING_TASK_ID, getId());
+    newContext.put(CompactSegments.STORE_COMPACTION_STATE_KEY, getContextValue(Tasks.STORE_COMPACTION_STATE_KEY, true));
```
Comment (suneet-s) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:425]: I think you should use putIfAbsent. That will allow the user to provide a `STORE_COMPACTION_STATE_KEY` that is different from the default value of true if needed.

Also nit: use a constant for the default value

File: core/src/main/java/org/apache/druid/timeline/DataSegment.java (hunk 1 above)
Comment (maytasm) [reply core/src/main/java/org/apache/druid/timeline/DataSegment.java:103]: Done

File: docs/querying/sql.md (hunk 2 above)
Comment (maytasm) [reply docs/querying/sql.md:1091]: Done. It's very basic example but any one with proficient SQL can probably think of better things to query.

File: sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java (hunk 3 above)
Comment (maytasm) [reply sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java:316]: Both `PartitionsSpec` and `indexSpec` can be helpful in determining the cause of query performance issue. We will also want to know what segments auto compaction will compact as both of those Spec determine that. 'indexSpec' can also impact query performance (see shape shifting PR). 

CompactionState is a small json and is already stored in memory as part of the DataSegment. We are also returning the shardSpec so this should not be a problem.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java (hunk 4 above)
Comment (maytasm) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java:425]: Done

File: docs/querying/sql.md [hunk 5]
```diff
@@ -1108,6 +1108,18 @@ GROUP BY 1
 ORDER BY 2 DESC
 ```
 
+If you want to retrieve segment that was compacted (ANY compaction):
+
+```sql
+SELECT * FROM sys.segments WHERE last_compaction_state is not null
+```
```
Comment (suneet-s) [inline docs/querying/sql.md:1115]: Suggested re-write to talk about the problem an admin would be looking to solve more explicitly.

```suggestion
If you want to know whether segments in the wikipedia table are not compacted, and therefore may be contributing to query slowness, you can run a query like this to tell you how many segments have not been compacted per interval.

```sql
SELECT start, end, count(*) FROM sys.segments WHERE datasource = 'wikipedia' and last_compaction_state is not null group by 1, 2
```

If you find that queries that hit these intervals are slow - consider enabling [compaction](../operations/segment-optimization.md) to produce more optimal segments.
```

File: docs/querying/sql.md [hunk 6]
```diff
@@ -1108,6 +1108,18 @@ GROUP BY 1
...
+
+or if you want to retrieve segment that was compacted only by a particular compaction spec (such as that of the auto compaction):
```
Comment (suneet-s) [inline docs/querying/sql.md:1117]:
```suggestion
Sometimes, you may have updated your compaction spec to a more optimal spec. In these cases you can find which segments are using the latest compaction spec using a query like the one below:
```

Can you also add instructions on how to produce the `last_compaction_state` blob from the web console. If I'm looking at the compaction config for a datasource - can I copy something to paste in the where clause here?

**Review by suneet-s — APPROVED** [review]
LGTM - some suggested re-wording, and a question about how to generate the last_compaction_state without actually looking at the values in the sys.segments table first.

No need to worry about my suggested re-wordings - that part is ok as is, I just thought I'd offer something different if you wanted to consider

File: docs/querying/sql.md (hunk 5 above)
Comment (maytasm) [reply docs/querying/sql.md:1115]: Done

File: docs/querying/sql.md (hunk 6 above)
Comment (maytasm) [reply docs/querying/sql.md:1117]: Done

**Comment by maytasm** [comment]
I'll make the change with re-write to docs in a separate PR.

**Pull request merged**